src/v/io/io_queue.cc

```diff
@@ -0,0 +1,197 @@
+
+    /*
+     * if it isn't queued, then that means the dispatcher has handled it. in
+     * this case, new data may have arrived into the page. we mark it as queued
+     * as a signal to the I/O completion continuation to requeue the write.
+     */
+    if (!page.test_flag(page::flags::queued)) {
+        log.debug("Marking inflight I/O to be requeued");
+        page.set_flag(page::flags::queued);
+    }
+}
+
+void io_queue::enqueue_pending(page& page) noexcept {
+    pending_.push_back(page);
+    page.set_flag(page::flags::queued);
+    cond_.signal();
+}
+
+seastar::future<> io_queue::dispatch() noexcept {
+    while (true) {
+        co_await cond_.wait([this] { return !pending_.empty() || stop_; });
+
+        if (stop_) {
+            co_return;
+        }
+
+        auto units = co_await seastar::get_units(ops_, 1);
+
+        if (file_ == nullptr) {
+            continue;
+        }
+
+        auto& page = pending_.front();
+        pending_.pop_front();
+        assert(page.test_flag(page::flags::queued));
+        page.clear_flag(page::flags::queued);
+        running_.push_back(page);
+
+        if (page.test_flag(page::flags::read)) {
+            dispatch_read(page, std::move(units));
+
+        } else if (page.test_flag(page::flags::write)) {
+            dispatch_write(page, std::move(units));
+
+        } else {
+            assert(false);
+            std::terminate();
+        }
+    }
+}
+
+void io_queue::dispatch_read(
+  page& page, seastar::semaphore_units<> units) noexcept {
+    log.debug("Reading {} at {}~{}", path_, page.offset(), page.data().size());
+
+    std::ignore = seastar::with_gate(
+      gate_, [this, &page, units = std::move(units)]() mutable {
+          return file_
+            ->dma_read(
+              page.offset(), page.data().get_write(), page.data().size())
+            .then([this, &page, units = std::move(units)](auto) {
+                running_.erase(request_list_type::s_iterator_to(page));
+                if (complete_) {
+                    complete_(page);
+                }
+            })
+            .handle_exception([this](std::exception_ptr eptr) {
+                log.error("Read failed to {}: {}", path_, eptr);
+                std::terminate();
+            });
+      });
+}
+
+void io_queue::dispatch_write(
+  page& page, seastar::semaphore_units<> units) noexcept {
+    log.debug("Writing {} at {}~{}", path_, page.offset(), page.data().size());
+
+    std::ignore = seastar::with_gate(
+      gate_, [this, &page, units = std::move(units)]() mutable {
+          return file_
+            ->dma_write(page.offset(), page.data().get(), page.data().size())
+            .then([this, &page, units = std::move(units)](auto) {
+                running_.erase(request_list_type::s_iterator_to(page));
+                if (page.test_flag(page::flags::queued)) {
+                    pending_.push_back(page);
+                    cond_.signal();
+                }
```

Comment:
Question: it looks like we clear the queued bit before calling `dispatch_write`. When will this code path be taken? 

Why doesn't a similar code path exist in `dispatch_read`?

Reply:
Great question. Added comments explaining this:

```diff
commit 049eaf3b462551d89e879dfbfe7b51fffd413107 (HEAD -> io-updates)
Author: Noah Watkins <noahwatkins@gmail.com>
Date:   Tue Feb 6 12:12:14 2024 -0800

    io: add comment explaining read/write re-queueing

    Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

diff --git a/src/v/io/io_queue.cc b/src/v/io/io_queue.cc
index f78210015e..6bf1bc477a 100644
--- a/src/v/io/io_queue.cc
+++ b/src/v/io/io_queue.cc
@@ -157,6 +157,10 @@ void io_queue::dispatch_read(
             ->dma_read(
               page.offset(), page.data().get_write(), page.data().size())
             .then([this, &page, units = std::move(units)](auto) {
+                /*
+                 * unlike writes, reads are not requeued. the caller should
+                 * handle read/write coherency.
+                 */
                 running_.erase(request_list_type::s_iterator_to(page));
                 if (complete_) {
                     complete_(page);
@@ -179,6 +183,12 @@ void io_queue::dispatch_write(
             ->dma_write(page.offset(), page.data().get(), page.data().size())
             .then([this, &page, units = std::move(units)](auto) {
                 running_.erase(request_list_type::s_iterator_to(page));
+                /*
+                 * the queued flag is cleared before dispatch_write is invoked,
+                 * but while the write is in flight, submit_write may be called
+                 * again. for example, if more data was written into the page
+                 * and the page needs to be written again. it's requeued here.
+                 */
                 if (page.test_flag(page::flags::queued)) {
                     pending_.push_back(page);
                     cond_.signal();
```

Reply:
Thanks for adding an explanation, this makes sense. 

A queued write request for a page is put into the `pending_` queue in continuation after `dma_write`, instead of doing this within `submit_write` when the request is first seen. Is this done to ensure some sort of serialization of writes? 

Or is it because we cannot keep the page in both pending and running intrusive lists at the same time?

Reply:
> A queued write request for a page is put into the pending_ queue in continuation after dma_write, instead of doing this within submit_write when the request is first seen. 

It is also added to the pending list in submit_write (via enqueue_pending). See here:
https://github.com/redpanda-data/redpanda/pull/16454/files#diff-7658d766d4b0e6184a27d049ab169d7b8ad7e53e56e478c6c8b345db7e15559eR96

> Is this done to ensure some sort of serialization of writes?

There is no ordering guarantees. Submitted I/Os can be serviced in any order.

> Or is it because we cannot keep the page in both pending and running intrusive lists at the same time?

This is true, there is only one intrusive list hook in the page.

More broadly, what is going on here is that pages are shared between the caller and the I/O queue. When data arrives into a page and the page is submitted to the I/O queue for write-back we need to be able to know if the newly arrived data will be covered by the _next_ dma_write or if a dma_write was racing with the submission, should the page be re-inserted for write-back so that the newly arrived data will also be written to disk.